src/config.rs

```diff
@@ -71,37 +131,108 @@ fn find_config() -> Option<PathBuf> {
     } else {
         None
     }
 }
 
+fn read_config_file(filename: &str) -> Result<String, Error> {
+    let path = find_config(filename).ok_or_else(|| {
+        Error::new(
+            ErrorKind::InvalidConfig,
+            format!("{} was not found in any location", filename),
+        )
+    })?;
+
+    read_to_string(path).map_err(|e| {
+        Error::new(
+            ErrorKind::InvalidConfig,
+            format!("Cannot read {}: {}", filename, e),
+        )
+    })
+}
+
 /// Create a `Session` from the config file.
 pub fn from_config<S: AsRef<str>>(cloud_name: S) -> Result<Session, Error> {
-    let path = find_config().ok_or_else(|| {
+    let mut clouds: serde_yaml::Value = serde_yaml::from_str(&read_config_file("clouds.yaml")?)
+        .map_err(|e| {
+            Error::new(
+                ErrorKind::InvalidConfig,
+                format!("Cannot parse clouds.yaml: {}", e),
+            )
+        })?;
+
+    let clouds_public: serde_yaml::Value = serde_yaml::from_str(
+        // If clouds-public.yaml is missing, let's pretend that it's there but empty.
+        &read_config_file("clouds-public.yaml").unwrap_or_else(|_| String::from("---\n{}\n")),
+    )
+    .map_err(|e| {
         Error::new(
             ErrorKind::InvalidConfig,
-            "clouds.yaml was not found in any location",
+            format!("Cannot parse clouds-public.yaml: {}", e),
         )
     })?;
-    let file = File::open(path).map_err(|e| {
+
+    let secure: serde_yaml::Value = serde_yaml::from_str(
+        // If secure.yaml is missing, let's pretend that it's there but empty.
+        &read_config_file("secure.yaml").unwrap_or_else(|_| String::from("---\n{}\n")),
+    )
+    .map_err(|e| {
         Error::new(
             ErrorKind::InvalidConfig,
-            format!("Cannot read config.yaml: {}", e),
+            format!("Cannot parse secure.yaml: {}", e),
         )
     })?;
-    let mut clouds_root: Root = serde_yaml::from_reader(file).map_err(|e| {
+
+    merge_mappings(
+        secure.as_mapping().ok_or_else(|| {
+            Error::new(
+                ErrorKind::InvalidConfig,
+                format!("secure.yaml's root is not a Mapping"),
+            )
+        })?,
+        clouds.as_mapping_mut().ok_or_else(|| {
+            Error::new(
+                ErrorKind::InvalidConfig,
+                format!("clouds.yaml's root is not a Mapping"),
+            )
+        })?,
+    );
+
+    clouds = inject_profiles(
+        clouds_public.as_mapping().ok_or_else(|| {
+            Error::new(
+                ErrorKind::InvalidConfig,
+                format!("clouds-public.yaml's root is not a Mapping"),
+            )
+        })?,
+        clouds.as_mapping_mut().ok_or_else(|| {
+            Error::new(
+                ErrorKind::InvalidConfig,
+                format!("clouds.yaml's root is not a Mapping"),
+            )
+        })?,
+    )?;
+
+    let clouds_string = serde_yaml::to_string(&clouds).map_err(|e| {
```

Comment:
Please use `serde_yaml::from_value` instead of converting to string and back.

Reply:
Done.